### ISSUE TITLE
Typescript 2.1.4 is unable to find node module

### ISSUE THREAD

**Issue by u-nikos** [Microsoft/TypeScript#12748]
**TypeScript Version:**  2.1.4

**Code**

```ts
$ echo 'import * as propertiesReader from "properties-reader";' > index.ts
$ npm install typescript@2.1.1 properties-reader
$ node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts
$ npm install typescript@2.1.4 properties-reader
$ node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts
index.ts(1,35): error TS2307: Cannot find module 'properties-reader'.
```

**Expected behavior:**
Should find the `properties-reader` node module.

**Actual behavior:**
Does not find the `properties-reader` node module.

**Comment by mhegazy** [comment]
I can not get this to repro locally..

Also on a side note, with the [changes for allowing untyped module imports](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#untyped-imports), i do not think you need `--allowJs --maxNodeModuleJsDepth 1`
 
```sh
c:\test\12748>echo 'import * as propertiesReader from "properties-reader";' > index.ts

c:\test\12748>type index.ts
'import * as propertiesReader from "properties-reader";'

c:\test\12748>npm install typescript@2.1.4 properties-reader
test@1.0.0 c:\test\12748
+-- properties-reader@0.0.15
`-- typescript@2.1.4

npm WARN test@1.0.0 No description
npm WARN test@1.0.0 No repository field.

c:\test\12748>.\node_modules\.bin\tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts

c:\test\12748>echo %ERRORLEVEL%
0
```

**Comment by u-nikos** [comment]
Maybe u can reproduce it in a docker container?

Dockerfile:
```
FROM node:7
RUN echo 'import * as propertiesReader from "properties-reader";' > index.ts
RUN npm install typescript@2.1.1 properties-reader@0.0.15
RUN node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts
RUN npm install typescript@2.1.4 properties-reader@0.0.15
RUN node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts
```

Command:
```bash
docker build -t typescript-issue-12748 .
```

Output:
```
Sending build context to Docker daemon 22.53 MB
Step 1 : FROM node:7
 ---> 36dc1bb7a52b
Step 2 : RUN echo 'import * as propertiesReader from "properties-reader";' > index.ts
 ---> Using cache
 ---> 13ad9350af4b
Step 3 : RUN npm install typescript@2.1.1 properties-reader@0.0.15
 ---> Running in 0fd9bf5ce6d4
npm info it worked if it ends with ok
npm info using npm@3.10.10
npm info using node@v7.2.1
npm info attempt registry request try #1 at 8:21:01 AM
npm http request GET https://registry.npmjs.org/typescript
npm info attempt registry request try #1 at 8:21:01 AM
npm http request GET https://registry.npmjs.org/properties-reader
npm http 200 https://registry.npmjs.org/properties-reader
npm http 200 https://registry.npmjs.org/typescript
npm info retry fetch attempt 1 at 8:21:06 AM
npm info attempt registry request try #1 at 8:21:06 AM
npm http fetch GET https://registry.npmjs.org/properties-reader/-/properties-reader-0.0.15.tgz
npm info retry fetch attempt 1 at 8:21:06 AM
npm info attempt registry request try #1 at 8:21:06 AM
npm http fetch GET https://registry.npmjs.org/typescript/-/typescript-2.1.1.tgz
npm http fetch 200 https://registry.npmjs.org/properties-reader/-/properties-reader-0.0.15.tgz
npm http fetch 200 https://registry.npmjs.org/typescript/-/typescript-2.1.1.tgz
npm info attempt registry request try #1 at 8:21:07 AM
npm http request GET https://registry.npmjs.org/mkdirp
npm http 200 https://registry.npmjs.org/mkdirp
npm info retry fetch attempt 1 at 8:21:07 AM
npm info attempt registry request try #1 at 8:21:07 AM
npm http fetch GET https://registry.npmjs.org/mkdirp/-/mkdirp-0.3.5.tgz
npm http fetch 200 https://registry.npmjs.org/mkdirp/-/mkdirp-0.3.5.tgz
npm info lifecycle mkdirp@0.3.5~preinstall: mkdirp@0.3.5
npm info lifecycle properties-reader@0.0.15~preinstall: properties-reader@0.0.15
npm info lifecycle typescript@2.1.1~preinstall: typescript@2.1.1
npm info linkStuff mkdirp@0.3.5
npm info linkStuff properties-reader@0.0.15
npm info linkStuff typescript@2.1.1
npm info lifecycle mkdirp@0.3.5~install: mkdirp@0.3.5
npm info lifecycle properties-reader@0.0.15~install: properties-reader@0.0.15
npm info lifecycle typescript@2.1.1~install: typescript@2.1.1
npm info lifecycle mkdirp@0.3.5~postinstall: mkdirp@0.3.5
npm info lifecycle properties-reader@0.0.15~postinstall: properties-reader@0.0.15
npm info lifecycle typescript@2.1.1~postinstall: typescript@2.1.1
/
+-- properties-reader@0.0.15
| `-- mkdirp@0.3.5
`-- typescript@2.1.1

npm WARN enoent ENOENT: no such file or directory, open '/package.json'
npm WARN !invalid#1 No description
npm WARN !invalid#1 No repository field.
npm WARN !invalid#1 No README data
npm WARN !invalid#1 No license field.
npm info ok
 ---> b72234b5212e
Removing intermediate container 0fd9bf5ce6d4
Step 4 : RUN node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts
 ---> Running in 3b65b7097f8e
 ---> c7b6c6c91ebc
Removing intermediate container 3b65b7097f8e
Step 5 : RUN npm install typescript@2.1.4 properties-reader@0.0.15
 ---> Running in 8a3d914cddc1
npm info it worked if it ends with ok
npm info using npm@3.10.10
npm info using node@v7.2.1
npm info retry fetch attempt 1 at 8:21:13 AM
npm info attempt registry request try #1 at 8:21:13 AM
npm http fetch GET https://registry.npmjs.org/typescript/-/typescript-2.1.4.tgz
npm http fetch 200 https://registry.npmjs.org/typescript/-/typescript-2.1.4.tgz
npm info attempt registry request try #1 at 8:21:20 AM
npm http request GET https://registry.npmjs.org/mkdirp
npm http 304 https://registry.npmjs.org/mkdirp
npm info lifecycle properties-reader@0.0.15~preinstall: properties-reader@0.0.15
npm info lifecycle typescript@2.1.4~preinstall: typescript@2.1.4
npm info lifecycle typescript@2.1.1~preuninstall: typescript@2.1.1
npm info lifecycle typescript@2.1.1~uninstall: typescript@2.1.1
npm info lifecycle typescript@2.1.1~postuninstall: typescript@2.1.1
npm info lifecycle properties-reader@0.0.15~preuninstall: properties-reader@0.0.15
npm info lifecycle properties-reader@0.0.15~uninstall: properties-reader@0.0.15
npm info lifecycle properties-reader@0.0.15~postuninstall: properties-reader@0.0.15
npm info linkStuff properties-reader@0.0.15
npm info linkStuff typescript@2.1.4
npm info lifecycle properties-reader@0.0.15~install: properties-reader@0.0.15
npm info lifecycle typescript@2.1.4~install: typescript@2.1.4
npm info lifecycle properties-reader@0.0.15~postinstall: properties-reader@0.0.15
npm info lifecycle typescript@2.1.4~postinstall: typescript@2.1.4
/
+-- properties-reader@0.0.15
`-- typescript@2.1.4

npm WARN enoent ENOENT: no such file or directory, open '/package.json'
npm WARN !invalid#1 No description
npm WARN !invalid#1 No repository field.
npm WARN !invalid#1 No README data
npm WARN !invalid#1 No license field.
npm info ok
 ---> 308a9e4f6d68
Removing intermediate container 8a3d914cddc1
Step 6 : RUN node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts
 ---> Running in 7cec5e74dfc8
index.ts(1,35): error TS2307: Cannot find module 'properties-reader'.
The command '/bin/sh -c node ./node_modules/.bin/tsc --allowJs --moduleResolution node --maxNodeModuleJsDepth 1 index.ts' returned a non-zero code: 2
```

**Comment by BigDataSamuli** [comment]
@mhegazy I think your index.ts only contains the string 'import * as propertiesReader from "properties-reader";' i.e. not an actual import statement.

I have the same issue, seems to happen with packages that don't contain and index.js file in the root package folder.

**Comment by mkulke** [comment]
Actually i think this is a typescript issue. Among other lookup methods it's parsing `package.json` of a package and then verifies the existence of the referenced file in there. Sometimes the main script is referenced like this (which is legal, because the `main` field should behave like a node `require`):

```
...
"main": "src/PropertiesReader",
...
```

Typescript doesn't find a file, because it does not add a `.js` suffix when looking this file up. It does however add `ts/.d` suffixes, so i guess it's a bug - because it does not detect the file, i'll cook up a PR to fix this.

**Comment by BigDataSamuli** [comment]
@mkulke You are probably on the right track. I had this issue with a package called enzyme, and that one seems to have directory path ("build") as the value of the main field, so you'd need to account for that as well.

**Comment by mkulke** [comment]
@BigDataSamuli this would also be fixable, but require some more profound changes to the module importer code.

**Comment by mhegazy** [comment]
> Actually i think this is a typescript issue. Among other lookup methods it's parsing package.json of a package and then verifies the existence of the referenced file in there. Sometimes the main script is referenced like this (which is legal, because the main field should behave like a node require):

@mkulke this issue is tracked by https://github.com/Microsoft/TypeScript/issues/12921

**Comment by mhegazy** [comment]
@u-nikos looks like the docker issue is a different issue. @andy-ms is looking into another docker issue as well.

**Comment by mhegazy** [comment]
the issue reported by @mkulke should be fixed by #13105. the issue OP seems to be a duplicate of https://github.com/Microsoft/TypeScript/issues/13026 and https://github.com/Microsoft/TypeScript/issues/13060. We should have a fix for this out soon.
